<a href="https://colab.research.google.com/github/jiangenhe/insc-486-fall-2021/blob/main/assignment5/Assignment_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# INSC486 Assignment 5 

- Submit your Colab link through Canvas
- This assignment is worth 90 points that accounts for 9% of your final grade.
- **Code quality will be considered in grading.**

Learning Objectives:
- Logistic regression (Statistical analysis)
- Build a supervised machine learning model (Classification)
- Enhance your pandas skills

## Task 1: Poisson regression analysis (10 points)

In [ ]:
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.api as sm
import numpy as np

**Data**: https://raw.githubusercontent.com/jiangenhe/insc-486-fall-2021/main/assignment5/poisson_sim.csv

**num_awards** is the outcome variable and indicates the number of awards earned by students at a high school in a year, **math** is a continuous predictor variable and represents students’ scores on their math final exam, and **prog** is a categorical predictor variable with three levels indicating the type of program in which the students were enrolled. It is coded as 1 = “General”, 2 = “Academic” and 3 = “Vocational”. Let’s start with loading the data and looking at some descriptive statistics.

### 1. Perform a Poisson regression analaysis to predict the number of awards earned by students at one high school. 

Predictors of the number of awards earned include the type of program in which the student was enrolled (e.g., vocational, general or academic) and the score on their final exam in math.

### 2. Interpret the regression results.

## Task 2: Build a classifier (40 points)

For this task, you will be using the Breast Cancer Wisconsin (Diagnostic) Database to create a classifier that can help diagnose patients. First, read through the description of the dataset (below).

In [2]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_breast_cancer

cancer = load_breast_cancer()

print(cancer.DESCR) # Print the data set description

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

The object returned by load_breast_cancer() is a scikit-learn Bunch object, which is similar to a dictionary.

In [3]:
cancer.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

### 1. Convert the sklearn.dataset `cancer` to a `(569, 31)` DataFrame `cancerdf`. (Code for this question has been provided)

In [16]:
cancerdf = pd.concat([pd.DataFrame(cancer.data ), pd.DataFrame(cancer.target)], axis=1)
cancerdf.describe()

0           1           2   ...          28          29          0 
count  569.000000  569.000000  569.000000  ...  569.000000  569.000000  569.000000
mean    14.127292   19.289649   91.969033  ...    0.290076    0.083946    0.627417
std      3.524049    4.301036   24.298981  ...    0.061867    0.018061    0.483918
min      6.981000    9.710000   43.790000  ...    0.156500    0.055040    0.000000
25%     11.700000   16.170000   75.170000  ...    0.250400    0.071460    0.000000
50%     13.370000   18.840000   86.240000  ...    0.282200    0.080040    1.000000
75%     15.780000   21.800000  104.100000  ...    0.317900    0.092080    1.000000
max     28.110000   39.280000  188.500000  ...    0.663800    0.207500    1.000000

[8 rows x 31 columns]

### 2. What is the class distribution? (i.e. how many instances of `malignant` (encoded 0) and how many `benign` (encoded 1)?)


### 3. Split the DataFrame into `X` (the data) and `y` (the labels).

* `X`*, a pandas DataFrame, has shape* `(569, 30)`
* `y`*, a pandas Series, has shape* `(569,)`.

### 4. Using `train_test_split`, split `X` and `y` into training and test sets `X_train`, `X_test`, `y_train`, and `y_test`.

* `X_train` *has shape* `(426, 30)`
* `X_test` *has shape* `(143, 30)`
* `y_train` *has shape* `(426,)`
* `y_test` *has shape* `(143,)`

In [ ]:
from sklearn.model_selection import train_test_split


### 5. Using KNeighborsClassifier, fit a k-nearest neighbors (knn) classifier with `X_train`, `y_train` and using one nearest neighbor (`n_neighbors = 1`).

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

knn = 

### 6. Using your knn classifier, predict the class label using the mean value for each feature.

Hint: You can use `cancerdf.mean()[:-1].values.reshape(1, -1)` which gets the mean value for each feature, ignores the target column, and reshapes the data from 1 dimension to 2 (necessary for the precict method of KNeighborsClassifier).

### 7. Use the knn classifer your created for Question 5 to predict the class labels for the test set `X_text`.

*The result should be a numpy array with shape `(143,)` and values either `0.0` or `1.0`.*

### 8. Find the score (mean accuracy) of your knn classifier using `X_test` and `y_test`.

*The calculated score should be a float between 0 and 1*

### 9. Read and run the code below to visualize the differet predicition scores between training and test sets, as well as malignant and benign cells.

*Correct variable names if they are different from yours, including `X_train`, `y_train`, `X_test`, `y_test` from question 4 and `knn` from question 5.*

In [ ]:
import matplotlib.pyplot as plt
%matplotlib notebook

# Find the training and testing accuracies by target value (i.e. malignant, benign)
mal_train_X = X_train[y_train==0]
mal_train_y = y_train[y_train==0]
ben_train_X = X_train[y_train==1]
ben_train_y = y_train[y_train==1]

mal_test_X = X_test[y_test==0]
mal_test_y = y_test[y_test==0]
ben_test_X = X_test[y_test==1]
ben_test_y = y_test[y_test==1]


scores = [knn.score(mal_train_X, mal_train_y), knn.score(ben_train_X, ben_train_y), knn.score(mal_test_X, mal_test_y), knn.score(ben_test_X, ben_test_y)]

plt.figure()

# Plot the scores as a bar chart
bars = plt.bar(np.arange(4), scores, color=['#4c72b0','#4c72b0','#55a868','#55a868'])

# directly label the score onto the bars
for bar in bars:
  height = bar.get_height()
  plt.gca().text(bar.get_x() + bar.get_width()/2, height*.90, '{0:.{1}f}'.format(height, 2), ha='center', color='w', fontsize=11)

# remove all the ticks (both axes), and tick labels on the Y axis
plt.tick_params(top='off', bottom='off', left='off', right='off', labelleft='off', labelbottom='on')

# remove the frame of the chart
for spine in plt.gca().spines.values():
  spine.set_visible(False)

plt.xticks([0,1,2,3], ['Malignant\nTraining', 'Benign\nTraining', 'Malignant\nTest', 'Benign\nTest'], alpha=0.8);
plt.title('Training and Test Accuracies for Malignant and Benign Cells', alpha=0.8)

## Task 3: Classification Playspace (40%)

For this task, you will build classifiers to classify an iris species as either (virginica, setosa, or versicolor) for the iris flower dataset. You will use different machine learning algorithm to build the classifiers.

### 0. Load Data

In [ ]:
import seaborn as sns
iris = sns.load_dataset("iris")

# X = feature values, all the columns except the last column
X = iris.iloc[:, :-1]

# y = target values, last column of the data frame
y = iris.iloc[:, -1]

### 1. Build a Basic Understanding of the dataset

Use head(), describe() and any other pandas features to explore the dataset.

### 2. Visualize the Dataset
Read the code and run it. You don't need to make any change.

In [ ]:
import matplotlib.pyplot as plt

plt.xlabel('Features')
plt.ylabel('Species')

pltX = iris.loc[:, 'sepal_length']
pltY = iris.loc[:,'species']
plt.scatter(pltX, pltY, color='blue', label='sepal_length')

pltX = iris.loc[:, 'sepal_width']
pltY = iris.loc[:,'species']
plt.scatter(pltX, pltY, color='green', label='sepal_width')

pltX = iris.loc[:, 'petal_length']
pltY = iris.loc[:,'species']
plt.scatter(pltX, pltY, color='red', label='petal_length')

pltX = iris.loc[:, 'petal_width']
pltY = iris.loc[:,'species']
plt.scatter(pltX, pltY, color='black', label='petal_width')

plt.legend(loc=4, prop={'size':8})
plt.show()


**For question 3 - 6, you are feel to set hyperparameters that you think is reasonable.**

### 3. Use kNN to build a classifier, including train/test data split, fit, and evaluation.

### 4. Use logistic model to build a classifier, including train/test data split, fit, and evaluation.

### 5. Use SVM to build a classifier, including train/test data split, fit, and evaluation.

### 6. Use Decision Tree Model to build a classifier, including train/test data split, fit, and evaluation.

### 7. Select one of the models to set different hyperparameters to compare the performace difference.

### 8. Select one of the models to apply cross validataion.